image.go

```diff
@@ -393,15 +393,23 @@ func (i *containerImageRef) NewImageSource(ctx context.Context, sc *types.System
 			rc.Close()
 			return nil, errors.Wrapf(err, "error opening file for %s", what)
 		}
-		destHasher := digest.Canonical.Digester()
+
 		counter := ioutils.NewWriteCounter(layerFile)
-		multiWriter := io.MultiWriter(counter, destHasher.Hash())
-		// Compress the layer, if we're recompressing it.
-		writeCloser, err := archive.CompressStream(multiWriter, i.compression)
-		if err != nil {
-			layerFile.Close()
-			rc.Close()
-			return nil, errors.Wrapf(err, "error compressing %s", what)
+		var destHasher digest.Digester
+		var writeCloser io.WriteCloser
+		if i.compression != archive.Uncompressed {
+			destHasher = digest.Canonical.Digester()
+			multiWriter := io.MultiWriter(counter, destHasher.Hash())
+			// Compress the layer, if we're recompressing it.
+			writeCloser, err = archive.CompressStream(multiWriter, i.compression)
+			if err != nil {
+				layerFile.Close()
+				rc.Close()
+				return nil, errors.Wrapf(err, "error compressing %s", what)
+			}
+		} else {
+			destHasher = srcHasher
+			writeCloser = ioutils.NopWriteCloser(counter)
```

Comment:
FWIW, I think we can move the `layerFile.Close()` line from line 442 up to line 413.5, just after this else.

Reply:
I don’t think so — we need the chain of writers open until `io.Copy` at line 440 actually writes the data.